src/jit/compiler.hpp

```diff
@@ -3680,7 +3680,11 @@ inline bool Compiler::IsGcSafePoint(GenTree* tree)
         GenTreeCall* call = tree->AsCall();
         if (!call->IsFastTailCall())
         {
-            if (call->gtCallType == CT_INDIRECT)
+            if (call->IsUnmanaged() && call->IsSuppressGCTransition())
```

Comment:
Is this still needed now that we are emitting the GC probe after the call?

Reply:
Nope. This can be removed.

Reply:
I think this actually is needed. The GC poll is not being inserted now.

Reply:
I though that we will insert GC poll unconditionally. Is it not the case?

Reply:
No, I consume this call in `morph.cpp` (see below) and this call is now returning `true` that the call is a GC safe point. I am out of my depth on this one. I will need to debug and understand the morph step a bit more. The branch is needed since `fgCreateGCPoll()` asserts that the basic block isn't a GC safe point so something needs to change here.

https://github.com/dotnet/coreclr/blob/7f3d58695ac3019d46059c11af2043b8a22a70fa/src/jit/morph.cpp#L8260-L8269

Reply:
This is not something that I'm terribly familiar with either, but if I understand correctly, you're saying that `ISGcSafePoint(call)` is now returning true? It looks like you removed the change to that method; perhaps what is needed is the reverse, that is (in `Compiler::IsGcSafePoint`):
```
            if ((call->gtCallType == CT_INDIRECT) && !(call->IsUnmanaged() && call->IsSuppressGCTransition()))
            {
                return true;
            }
```
Or something like that. Perhaps some one else on the JIT team is more familiar with this. @dotnet/jit-contrib

Reply:
@CarolEidt I did remove it because I thought it was going to return `false`. Adding it back does make it work as expected, but given my current confusion here I am inclined to debug through this again and verify I actually understand the reasoning for why it is returning `true` and when and why the various flags are set.

Any additional insight in this area is welcome.

Reply:
@CarolEidt Thanks for the suggestion. I verified the various options and I believe your suggestion is what I am going to do. I will follow the branch logic in the `CT_USER_FUNC` case and create an inner predicate check instead of chaining them together, but the logic is the same. Thanks!

Reply:
Bah. Missed the `CT_USER_FUNC` branch for an inline P/Invoke. Looks like I am still missing something.

Reply:
Okay, the last check is for `GTF_CALL_M_NOGCCHECK`. This is set for FCalls, which seems like something I could be reusing instead of creating the new `GTF_CALL_M_SUPPRESS_GC_TRANSITION`. From the standpoint of this feature an FCall is what we are turning a P/Invoke into so it might make sense to reuse that flag. Opinions?

compiler.hpp:
```C++
if (call->gtCallType == CT_INDIRECT)
{
    if (!(call->IsUnmanaged() && call->IsSuppressGCTransition()))
    {
        return true;
    }
}
else if (call->gtCallType == CT_USER_FUNC)
{
    if ((call->gtCallMoreFlags & GTF_CALL_M_NOGCCHECK) == 0)
    {
        return true;
    }
}
```

jitinterface.cpp:
``` C++
if (pMD->IsFCallOrIntrinsic())
    result |= CORINFO_FLG_NOGCCHECK | CORINFO_FLG_INTRINSIC;
```

Reply:
Oh that's right. I forgot I had rejected this approach because I don't want to insert a GCPoll in any FCalls. Going to need to keep this and simply make the same check in the `CT_USER_FUNC` branch.

Reply:
This is pretty convoluted, and could use some additional documentation, at least in compiler.hpp, if not elsewhere to distinguish between `GTF_CALL_M_NOGCCHECK` and `GTF_CALL_M_SUPPRESS_GC_TRANSITION`. Is it correct to say that the former says "this is a call that does not require a GC check (i.e. it is inherently a GC safe point), but the latter says "this call is suppressing the GC transition, but requires a separate GC safe point"?

Reply:
@CarolEidt That is nicely worded, yes. This is implied in `corinfo.h` - see below.

https://github.com/dotnet/coreclr/blob/7f3d58695ac3019d46059c11af2043b8a22a70fa/src/inc/corinfo.h#L831

The above documentation states this is used for an FCall - which by definition doesn't have the GC check and even provides the helpful hint of "Don't put alone in loops". The corresponding documentation for the JIT flag is a bit less helpful.

https://github.com/dotnet/coreclr/blob/7f3d58695ac3019d46059c11af2043b8a22a70fa/src/jit/gentree.h#L3694

The new flag for the JIT is documented as such:

https://github.com/dotnet/coreclr/blob/7f3d58695ac3019d46059c11af2043b8a22a70fa/src/jit/gentree.h#L3734

I agree additional context should be given with this usage or on the flag. What are your thoughts here? At the usage point or perhaps where the JIT flag is defined? I am updating this code to what I originally had plus a comment on why. I would concur the usage of the `` flag in this specific call might also warrant a specific comment but it could instead be placed on the flag definition itself. I am fine doing whatever is most helpful.

```c++
if (!call->IsFastTailCall())
{
    if (call->IsUnmanaged() && call->IsSuppressGCTransition())
    {
        // Both an indirect and user func call can be unmanaged
        // and have a request to suppress the GC transition so
        // the check is pulled out into its own branch.
        return false;
    }
    else if (call->gtCallType == CT_INDIRECT)
    {
        return true;
    }
    else if (call->gtCallType == CT_USER_FUNC)
    {
        if ((call->gtCallMoreFlags & GTF_CALL_M_NOGCCHECK) == 0)
        {
            return true;
        }
    }
    // otherwise we have a CT_HELPER
}
```

Reply:
It is the case, but we need to detect when we are in a basic block that isn't already marked as a GC safe point.